api.go

```diff
@@ -0,0 +1,56 @@
+package sqlparser
+
+// Visit defines the signature of a function that
+// can be used to visit all nodes of a parse tree.
+type Visit func(node Node) (kontinue bool, err error)
+
+// Walk calls visit on every node.
+// If visit returns true, the underlying nodes
+// are also visited. If it returns an error, walking
+// is interrupted, and the error is returned.
+func Walk(visit Visit, nodes ...Node) error {
+	for _, node := range nodes {
+		if node == nil {
+			continue
+		}
+		kontinue, err := visit(node)
+		if err != nil {
+			return err
+		}
+		if kontinue {
+			err = node.walkSubtree(visit)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// GetUniqueTableReferences returns a slice of tables' names referenced by the node.
+func GetUniqueTableReferences(node Node) []string {
+	if node == nil {
+		return []string{}
+	}
+
+	tables := map[string]struct{}{}
+	tableNames := []string{}
+	if table, ok := node.(*Table); ok && table != nil {
+		tables[table.Name.String()] = struct{}{}
+		tableNames = append(tableNames, table.Name.String())
+	}
+
+	// it's ok to ignore the error because the visit function does not throw an error
+	_ = node.walkSubtree(func(node Node) (bool, error) {
+		if table, ok := node.(*Table); ok && table != nil {
+			tableName := table.Name.String()
+			if _, ok := tables[tableName]; !ok {
+				tables[tableName] = struct{}{}
+				tableNames = append(tableNames, tableName)
+			}
+		}
+		return true, nil
+	})
```

Comment:
Couldn't we switch L36-L53 with a single `Walk(...,  node)` call?
That would allow avoiding what we need to do between L36-L41 to "manually" call the "visit" logic for the "top node".

Reply:
I thought a little about it but could not see a solution without creating another function. Do you see any?

Reply:
Well, in my comment I meant _literally_ replace L38-L53 with:
```
_ = Walk(func(node Node) (bool, error) {
		if table, ok := node.(*Table); ok && table != nil {
			tableName := table.Name.String()
			if _, ok := tables[tableName]; !ok {
				tables[tableName] = struct{}{}
				tableNames = append(tableNames, tableName)
			}
		}
		return true, nil
	}, node)
```

Wouldn't that work? Maybe I'm missing something.

Reply:
Cool, nice catch.